text/0000-bee-pow/0000-bee-pow.md

```diff
@@ -0,0 +1,332 @@
+}
+```
+
+Notive that by making `Data` and `Hash` associated types this RFC doesn't make any commitment to the underlying representation which allows for a very flexible design.
+
+## Sampler
+
+The `Sampler` trait allows implementing different types that mutate the input data handed to the hasher on each try. The standard sampler used in IOTA simply starts with a zero nonce and increments it until it finds a valid one. There are, however, other samplers possible, like a random sampler.
+
+```Rust
+pub trait Sampler {
+    type Data;
+
+    fn next(&mut self, data: &mut Self::Data);
+}
+```
+
+Like in the `Hasher` trait we now have to make `Data` an associated type as well, so it can always be compatible when used together in `PearlDiver`.
+
+## Tester
+
+The `Tester` trait allows implementing different types that place  plug different validation schemes into `PearlDiver`, like the `Hashcash` or the `Hamming` validation scheme.
+
+```Rust
+pub trait Tester {
+    type Hash;
+
+    fn is_valid(&self, hash: &Self::Hash) -> bool;
+}
+```
+
+As with `Sampler` the `Tester` trait needs an associated type so it can always be implemented in a compatible way and operate on something served by a type implementing the `Hasher` trait. 
+
+## PearlDiver
+
+With the traits above a type can now be defined that is called `PearlDiver`. It encapsulates the following algorithm:
+
+Starting with some serialized transaction data (which also encodes a zero nonce):
+1. hash the data
+2. check if the hash satisifies some pre-defined and adjustable constraint
+   1. hash satisfies constraint => stop the algorithm 
+   2. hash doesn't satisfy constraint 
+      1. select another nonce 
+      2. goto 1.
+
+This RFC proposes the following `PearlDiver` type:
+
+```Rust
+pub struct PearlDiver<H, S, T, D, R>
+where
+    H: Hasher<Data = D, Hash = R>,
+    S: Sampler<Data = D>,
+    T: Tester<Hash = R>,
+{
+    hasher: H,
+    sampler: S,
+    tester: T,
+}
+
+impl<H, S, T, D, R> PearlDiver<H, S, T, D, R>
+where
+    H: Hasher<Data = D, Hash = R>,
+    S: Sampler<Data = D>,
+    T: Tester<Hash = R>,
+{
+    pub fn new(hasher: H, sampler: S, tester: T) -> Self {
+        Self {
+            hasher,
+            sampler,
+            tester,
+        }
+    }
+
+    pub fn search(&mut self, mut data: &mut D) {
+        loop {
+            let hash = self.hasher.hash(data);
+            if self.tester.is_valid(&hash) {
+                break;
+            } else {
+                self.sampler.next(&mut data);
+            }
+        }
+    }
+}
+```
+## PoW
+
+Proof-of-Work essentially is "just" about finding a piece of data (the Nonce) that is - in some way - associated with the data at hand. This proposal therefore introduces a special `PoW` trait for that.
+
+```Rust
+pub trait PoW {
+    type Data;
+    type Nonce;
+
+    fn get_nonce(&mut self, data: Self::Data) -> PoWResult<Self::Nonce>);
+}
+```
+whereby
+
+```Rust
+pub type PoWError = Box<dyn std::error::Error>;
```

Comment:
Is there any particular reason that `PoWError` must be dynamically-allocated?

Reply:
I think not. Any suggestion?